### PR TITLE
fix(GiniCaptureSDK): BAC-894 send exif data

### PR DESCRIPTION
**Pull request by igor-gini**

### CODE REVIEW

**Review by ValentinaIancu-Gini — COMMENTED** [review]
@igor-gini Nice work.
Can you please check the suggestions that I left?

**Review by zladzeyka — CHANGES_REQUESTED** [review]
Thank you @igor-gini 🥳
Please find a small suggestion from my side
Could you please also add a Unit test for `metadata`?

File: BankAPILibrary/GiniBankAPILibrary/Sources/GiniBankAPILibrary/Documents/Document.swift [hunk 1]
```diff
@@ -237,23 +237,97 @@ extension Document {
...
+
         /**
          * The document metadata initializer with the branch ID (i.e: the BLZ of a Bank in Germany) and additional
          * headers.
          *
          * - Parameter branchId:            The branch id (i.e: the BLZ of a Bank in Germany)
          * - Parameter additionalHeaders:   Additional headers for the metadata. i.e: ["customerId":"123456"]
          */
-        public init(branchId: String? = nil, additionalHeaders: [String: String]? = nil) {
+        public init(branchId: String? = nil, uploadMetadata: UploadMetadata? = nil, additionalHeaders: [String: String]? = nil) {
```
[the remark this answers is not in the document]
Comment (igor-gini) [reply BankAPILibrary/GiniBankAPILibrary/Sources/GiniBankAPILibrary/Documents/Document.swift]: Sure, done

File: CaptureSDK/GiniCaptureSDK/Sources/GiniCaptureSDK/Core/Models/GiniCaptureDocument.swift [hunk 2]
```diff
@@ -102,6 +105,30 @@ public class GiniCaptureDocumentBuilder: NSObject {
             completion(self.build(with: data, fileName: openURL.lastPathComponent))
         }
     }
+
+    public func generateUploadMetadata() -> Document.UploadMetadata {
+        var deviceOrientation = ""
+        if let isLandscape = self.deviceOrientation?.isLandscape {
+            deviceOrientation = isLandscape ? "landscape" : "portrait"
+        }
```
[the remark this answers is not in the document]
Comment (igor-gini) [reply CaptureSDK/GiniCaptureSDK/Sources/GiniCaptureSDK/Core/Models/GiniCaptureDocument.swift]: Moved it to a separate file even

File: BankAPILibrary/GiniBankAPILibrary/Sources/GiniBankAPILibrary/Documents/Document.swift [hunk 3]
```diff
@@ -237,23 +237,97 @@ extension Document {
...
+                comment += ",\(userCommentEntryPoint)=\(entryPoint)"
+            }
+            if !importMethod.isEmpty {
+                comment += ",\(userCommentImportMethod)=\(importMethod)"
+            }
+            if !deviceOrientation.isEmpty {
+                comment += ",\(userCommentDeviceOrientation)=\(deviceOrientation)"
+            }
+            userComment = comment
+        }
```
[the remark this answers is not in the document]
Comment (igor-gini) [reply BankAPILibrary/GiniBankAPILibrary/Sources/GiniBankAPILibrary/Documents/Document.swift]: Sure

File: CaptureSDK/GiniCaptureSDK/Sources/GiniCaptureSDK/Core/Screens/Document picker/Gallery/GalleryCoordinator.swift [hunk 4]
```diff
@@ -293,10 +294,24 @@ extension GalleryCoordinator: ImagePickerViewControllerDelegate {
...
+            giniCaptureVersion: GiniCaptureSDKVersion,
+            deviceOrientation: "",
+            source: DocumentSource.external.value,
+            importMethod: DocumentImportMethod.picker.rawValue,
+            entryPoint: {
+                switch GiniConfiguration.shared.entryPoint {
+                case .button: "button"
+                case .field: "field"
+                }
+            }(),
```
[the remark this answers is not in the document]
Comment (igor-gini) [reply CaptureSDK/GiniCaptureSDK/Sources/GiniCaptureSDK/Core/Screens/Document picker/Gallery/GalleryCoordinator.swift]: Sure

File: CaptureSDK/GiniCaptureSDK/Sources/GiniCaptureSDK/Core/Screens/Camera/Camera/CameraButtonsViewModel.swift [hunk 5]
```diff
@@ -84,11 +85,25 @@ public final class CameraButtonsViewModel {
...
+            giniCaptureVersion: GiniCaptureSDKVersion,
+            deviceOrientation: orientation.isLandscape ? "landscape" : "portrait",
+            source: DocumentSource.camera.value,
+            importMethod: "",
+            entryPoint: {
+                switch GiniConfiguration.shared.entryPoint {
+                case .button: "button"
+                case .field: "field"
+                }
+            }(),
```
[the remark this answers is not in the document]
Comment (igor-gini) [reply CaptureSDK/GiniCaptureSDK/Sources/GiniCaptureSDK/Core/Screens/Camera/Camera/CameraButtonsViewModel.swift]: Done

File: CaptureSDK/GiniCaptureSDK/Sources/GiniCaptureSDK/Networking/GiniCaptureNetworkService.swift [hunk 6]
```diff
@@ -132,6 +132,14 @@ class DefaultCaptureNetworkService: GiniCaptureNetworkService {
     func upload(document: GiniCaptureDocument,
                 metadata: Document.Metadata?,
                 completion: @escaping UploadDocumentCompletion) {
+        var metadata: Document.Metadata? = metadata
+        if let uploadMetadata = document.uploadMetadata {
+            if metadata == nil {
```
[the remark this answers is not in the document]
Comment (igor-gini) [reply CaptureSDK/GiniCaptureSDK/Sources/GiniCaptureSDK/Networking/GiniCaptureNetworkService.swift]: Sure

File: CaptureSDK/GiniCaptureSDK/Sources/GiniCaptureSDK/Core/Helpers/ImageMetaInformationManager.swift [hunk 7]
```diff
@@ -232,43 +227,23 @@ final class ImageMetaInformationManager {
...
+            rotationNorm = "\(normalizedDegrees(imageRotation: imageRotationDeltaDegrees() + rotationDegrees))"
         }
-        return comment
+        return Document.UploadMetadata.constructComment(
+            osVersion: UIDevice.current.systemVersion,
+            giniVersion: GiniCapture.versionString,
+            contentId: imageUUID(),
+            source: imageSource.value,
+            entryPoint: GiniConfiguration.shared.entryPoint.stringValue,
+            importMethod: imageSource.value != DocumentSource.camera.value
```
Comment (ValentinaIancu-Gini) [inline CaptureSDK/GiniCaptureSDK/Sources/GiniCaptureSDK/Core/Helpers/ImageMetaInformationManager.swift:241]: [suggestion] Maybe for readability we can have let a property before return with the value for the importMethod. What do you say?

**Review by ValentinaIancu-Gini — COMMENTED** [review]
@igor-gini thank you for fixing my suggestions.  Can you please check the last ones?

File: CaptureSDK/GiniCaptureSDK/Sources/GiniCaptureSDK/Core/Helpers/ImageMetaInformationManager.swift [hunk 8]
```diff
@@ -92,13 +93,7 @@ final class ImageMetaInformationManager {
     // user comment fields
     let userCommentRotation = "RotDeltaDeg"
     let userCommentContentId = "ContentId"
-    let userCommentPlatform = "Platform"
-    let userCommentOSVer = "OSVer"
-    let userCommentGiniVersionVer = "GiniCaptureVer"
     let userCommentDeviceOrientation = "DeviceOrientation"
```
[the remark this answers is not in the document]
Comment (igor-gini) [reply CaptureSDK/GiniCaptureSDK/Sources/GiniCaptureSDK/Core/Helpers/ImageMetaInformationManager.swift:96]: Yes, those 3 are still needed, they seem to be used when extracting existing data from the EXIF

File: CaptureSDK/GiniCaptureSDK/Sources/GiniCaptureSDK/Core/Helpers/ImageMetaInformationManager.swift (hunk 7 above)
Comment (igor-gini) [reply CaptureSDK/GiniCaptureSDK/Sources/GiniCaptureSDK/Core/Helpers/ImageMetaInformationManager.swift]: Sure

File: BankAPILibrary/GiniBankAPILibrary/Sources/GiniBankAPILibrary/Documents/Document+UploadMetadata.swift [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+        var userComment: String
+        static let userCommentRotation = "RotDeltaDeg"
+        static let userCommentContentId = "ContentId"
+        static let userCommentPlatform = "Platform"
+        static let userCommentOSVer = "OSVer"
+        static let userCommentGiniVersionVer = "GiniCaptureVer"
+        static let userCommentDeviceOrientation = "DeviceOrientation"
+        static let userCommentSource = "Source"
+        static let userCommentImportMethod = "ImportMethod"
+        static let userCommentEntryPoint = "EntryPoint"
```
[the remark this answers is not in the document]
Comment (igor-gini) [reply BankAPILibrary/GiniBankAPILibrary/Sources/GiniBankAPILibrary/Documents/Document+UploadMetadata.swift]: Sure, done

File: BankAPILibrary/GiniBankAPILibrary/Tests/GiniBankAPILibraryTests/DocumentTests.swift [hunk 10]
```diff
@@ -111,6 +111,53 @@ final class GiniDocumentTests: XCTestCase {
...
+        let uploadMetadata = Document.UploadMetadata.init(
+            giniCaptureVersion: "99.99.99",
+            deviceOrientation: "deviceOrientation",
+            source: "source",
+            importMethod: "import",
+            entryPoint: "unit-test",
+            osVersion: "ios 99"
+        )
+        metadata.addUploadMetadata(uploadMetadata)
+        XCTAssertEqual(metadata.headers[Document.Metadata.headerKeyPrefix + Document.Metadata.uploadHeaderKey], uploadMetadata.userComment)
```
[the remark this answers is not in the document]
Comment (igor-gini) [reply BankAPILibrary/GiniBankAPILibrary/Tests/GiniBankAPILibraryTests/DocumentTests.swift]: Added

File: CaptureSDK/GiniCaptureSDK/Sources/GiniCaptureSDK/Core/Helpers/ImageMetaInformationManager.swift (hunk 8 above)
[the remark this answers is not in the document]
Comment (ValentinaIancu-Gini) [reply CaptureSDK/GiniCaptureSDK/Sources/GiniCaptureSDK/Core/Helpers/ImageMetaInformationManager.swift:96]: @igor-gini I think we can remove `userCommentDeviceOrientation`. I can not find it used anywhere in this file.

**Pull request closed without merging**